### ISSUE TITLE
add StatNameJoinPool

### ISSUE THREAD

**Issue by jmarantz** [envoyproxy/envoy#10516]
Currently `SymbolTable::join()` is cumbersome to call, as it returns a StoragePtr, from which you can construct a StatName. The StatName is valid as long as the StoragePtr is.

Those semantics are what they are, but I think all the call-sites for `SymbolTable::join` would be cleaner if their were a pool-like wrapper:
```
class StatNameJoinPool {
public:
  StatNameJoinPool(SymbolTable& symbol_table);
 `StatName join(const std::vector<StatName>& parts);
};
```
again the StatNameJoinPool would need to outlive the StatNames created with it, but I think this would be more intuitive at call-sites.

**Comment by jmarantz** [comment]
Note that at call-sites where a StoragePtr is allocated locally, and then a StatName is created around that, usage of this helper-pool does not reduce the # of lines of code, but maybe makes them a little more intuitive.

I'm reconsidering though calling it `StatNameJoinPool` as it does not actually manage the StatNames in the vector passed to join(), and calling it a `pool` might falsely imply that it would. It would be the caller's responsibility to ensure the StatNames passed to join() outlive the returned StatName.

**Comment by jmarantz** [comment]
This issue is resolved with #10735
